network/src/noise/handshake.rs

```diff
@@ -93,7 +94,7 @@ pub enum HandshakeAuthMode {
     /// In `ServerOnly` mode, the dialer authenticates the server. However, the
     /// server does not care who connects to them and will allow inbound connections
     /// from any peer.
-    ServerOnly,
+    MaybeMutual(Arc<RwLock<HashMap<PeerId, HashSet<x25519::PublicKey>>>>),
```

Comment:
hmm, do we need the anti-replay timestamps here as well?

Reply:
Well given the comment on it for the Mutual auth case I'm not convinced its what we want on public networks

Reply:
I'd say they wouldn't be bad but are not P0. I think in order to take advantage of them, you would /only/ use them for mutual authentication in the maybe mutual path. Without looking at the code, I don't know how complicated that would be, but I can imagine messing it up and creating a really ugly bug. I'm curious from @mimoo who probably remembers better if this is something we should consider prioritizing...

Reply:
also @phlip9

Reply:
Deleted my earlier comments, there are a few issues with timestamps and why they won't be that great for this case:
1) They have no real relationship with time, one could simply choose from 1 to infinity and the server will never verify they are within the current time window -- this could also have reduced or completely eliminated maintained state depending on the implementation
2) A maybe mutual auth already supports arbitrary peers pinging it, so even if we use replay prevention that would only prevent it on trusted peers as anyone can generate arbitrary keys (e.g., random bytes) to cause wasted DHs

So yeah, definitely not worth adding and maybe even worth fixing the comments

Reply:
If `trusted_peers` is always needed, it should probably be part of another data structure instead (like the state).
About the timestamp, the [network spec](https://github.com/diem/diem/blob/master/specifications/network/noise.md#replay-attacks) notes:

> We do not protect against this in the FN since an attacker can just create as many valid handshakes as they want.

Reply:
We could only do anti-replay detection for trusted peers, even on the public network. Then we can put off garbage collection for another day : )

Reply:
To clarify a bit, the threat here is a passive MitM that observes handshakes on the wire. Given DiemNet's current 1-connection 1-peer design, without replay attack detection, the MitM can evict honest peers' connections by replaying their handshakes. In lieu of more work implementing a more sophisticated `AntiReplayTimestamps` structure, we could just do anti replay detection for trusted peers.

Or maybe we bite the bullet and just do anti replay detection for everything. A simple approach might be for `AntiReplayTimestamps` to instead contain _two_ maps. One for trusted peers and one for untrusted peers. We also add a maximum capacity and just evict a random entry at capacity.